internal/handlers/common/aggregations/aggregations.go

```diff
@@ -24,6 +24,31 @@ import (
 
 // Stage is a common interface for all aggregation stages.
 type Stage interface {
+	// FetchDocuments fetches document iterator, it is only used if this is
+	// the first stage of the pipeline.
+	// The returned document iterator may be originated from querying database or
+	// from in-memory value iterator.
+	// This allows first stage of the pipeline to decide how to create initial document iterator.
+	FetchDocuments(ctx context.Context, closer *iterator.MultiCloser) (types.DocumentsIterator, error)
```

Comment:
Or maybe it's part of a "pipeline" interface. E.g. we "seed" the pipeline with some documents first, and then apply stages on it. And then it's pipeline responsibility/implementation  how to fetch these seeded documents from the DB.

Reply:
After some thinking... Actually, the current approach looks good to me too. I was confused by the phrase "it is only used if this is the first stage of the pipeline", but actually any stage could be the first stage, so it's correct that this method exists for each stage. 

On the other hand, the implementation of this method is exactly the same in most of the cases, so that's another sign to me that maybe this function shouldn't be here...

Reply:
This function seems reasonable if the way of fetching documents is different at the first stage for some stages, although I cannot define any of them

Reply:
Valid point about the implementation of this `FetchDocuments` being _mostly_ the same. For upcoming pipelines, the ones that have different implementations are `$documents`, `$listLocalSessions`, `$listSessions` and existing `$collStats`.

It cannot be part of pipeline because it would expose specific details about each stage to the pipeline. Let me try a few other things if it's possible to make it not an interface. Yes thanks we should rename this to something better :hugs:

Reply:
@chilagrow yeah, I agree with your point. My thought was to somehow "identify" the type of pipeline based on the first stage and then fetch documents based on it, but it doesn't seem better (because then we need to come up with a "pipeline" type which will depend on the first stage, so the implementation doesn't look clear and simple). 

So, it looks to me that the current function "How documents are fetched if this stage is the first in the pipeline" is a better solution.

Reply:
I tried your suggestion in another PR https://github.com/FerretDB/FerretDB/pull/3110
This approach exposes logic related to `$collStats` to aggregation package level. But it removes first stage logic from the stages. Let me know what you think :hugs: